File: python/tests/test_load_table.py
import duckdb
import pytest

from bedrock_bio.load_table import load_table
from conftest import requires_live_v2_manifest


class TestLoadTable:
    def test_errors_on_unknown_table(self):
        requires_live_v2_manifest()
        with pytest.raises(ValueError, match="not found in manifest"):
            load_table("not_a_table")

    def test_returns_relation(self):
        requires_live_v2_manifest()
        assert isinstance(load_table("dbsnp.vcf"), duckdb.DuckDBPyRelation)

    def test_filter_narrows(self):
        requires_live_v2_manifest()
        rel = (
            load_table("dbsnp.vcf")
            .filter("assembly = 'GRCh38' AND chromosome = '22'")
            .limit(5)
        )
        rows = rel.fetchall()
        assert len(rows) == 5
        chromosome = rel.columns.index("chromosome")
        assert {row[chromosome] for row in rows} == {"22"}

    def test_select_limits_columns(self):
        requires_live_v2_manifest()
        rel = (
            load_table("dbsnp.vcf")
            .filter("assembly = 'GRCh38' AND chromosome = '22'")
            .select("chromosome", "position")
            .limit(5)
        )
        assert rel.columns == ["chromosome", "position"]

    def test_filters_and_projection_push_down_to_scan(self):
        requires_live_v2_manifest()
        plan = (
            load_table("dbsnp.vcf")
            .filter("assembly = 'GRCh38' AND chromosome = '22' AND position > 50000000")
            .select("rsid, position")
            .explain()
        )
        assert "ICEBERG_SCAN" in plan
        assert "FILTER" not in plan
        assert "assembly='GRCh38'" in plan
        assert "chromosome='22'" in plan
        assert "position>50000000" in plan
        assert "ref_allele" not in plan
        assert "alt_allele" not in plan

    def test_reads_second_real_table(self):
        requires_live_v2_manifest()
        rows = load_table("ensembl.taxonomies").limit(3).fetchall()
        assert len(rows) == 3

File: r/tests/testthat/test-load_table.R
skip_on_cran()
skip_if_offline()

dbsnp <- function() {
  load_table("dbsnp.vcf") |>
    dplyr::filter(assembly == "GRCh38", chromosome == "22")
}

test_that("errors on unknown table", {
  skip_unless_live_v2()
  expect_error(
    load_table("not_a_table"),
    "not found in manifest"
  )
})

test_that("returns a lazy tbl", {
  skip_unless_live_v2()
  expect_s3_class(load_table("dbsnp.vcf"), "tbl_lazy")
})

test_that("filter narrows results", {
  skip_unless_live_v2()
  df <- head(dbsnp(), 5) |> dplyr::collect()
  expect_equal(nrow(df), 5L)
  expect_equal(unique(df$chromosome), "22")
})

test_that("select limits columns", {
  skip_unless_live_v2()
  df <- dbsnp() |>
    dplyr::select(chromosome, position) |>
    head(5) |>
    dplyr::collect()
  expect_equal(names(df), c("chromosome", "position"))
})

test_that("filters and projection are pushed into the scan", {
  skip_unless_live_v2()
  # Partition filters (assembly, chromosome), a non-partition predicate
  # (position), and a projection must all be pushed into the Iceberg scan
  # rather than applied after a full read -- the partition-pruning /
  # predicate-pushdown that makes the large tables usable.
  rel <- load_table("dbsnp.vcf") |>
    dplyr::filter(
      assembly == "GRCh38",
      chromosome == "22",
      position > 50000000L
    ) |>
    dplyr::select(rsid, position)
  con <- dbplyr::remote_con(rel)
  plan <- DBI::dbGetQuery(con, paste("EXPLAIN", dbplyr::sql_render(rel)))
  plan_text <- paste(unlist(plan), collapse = "\n")

  expect_match(plan_text, "ICEBERG_SCAN")
  # No standalone FILTER operator: all predicates reached the scan.
  expect_false(grepl("FILTER", plan_text))
  expect_match(plan_text, "assembly='GRCh38'")
  expect_match(plan_text, "chromosome='22'")
  expect_match(plan_text, "50000000")
  # Projection pushed down: unselected columns are pruned from the scan.
  expect_false(grepl("ref_allele", plan_text))
  expect_false(grepl("alt_allele", plan_text))
})

test_that("reads an unpartitioned table", {
  skip_unless_live_v2()
  # The other table shape: a single-file table with no partition columns
  # (dbsnp.vcf above is hive-partitioned).
  df <- load_table("ensembl.taxonomies") |>
    head(3) |>
    dplyr::collect()
  expect_equal(nrow(df), 3L)
  expect_true(ncol(df) > 0)
})
